File: Problems/1338_Reduce_Array_Size_to_The_Half.cpp
#include <vector>
#include <unordered_map>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minSetSize(vector<int>& arr) {
        unordered_map<int, int> hm;
        for (auto num : arr) {
            ++hm[num];
        }
        priority_queue <int> counts;
        for (auto it : hm) {
            counts.push(it.second);
        }

        int target = (arr.size() + 1) / 2;
        int res = 0;
        while (!counts.empty()) {
            int tmp = counts.top();
            counts.pop();
            target -= tmp;
            ++res;
            if (target <= 0) return res;
        }
        return res;
    }
};


class Solution {
public:
    int minSetSize(vector<int>& arr) {
        unordered_map<int, int> hm;
        for (auto num : arr) {
            ++hm[num];
        }
        
        vector<int> counts;
        for (auto it : hm) {
            counts.push_back(it.second);
        }
        sort(counts.begin(), counts.end());
        reverse(counts.rbegin(), counts.rend());

        int target = (arr.size() + 1) / 2;
        int res = 0;
        for (auto cnt : counts) {
            target -= cnt;
            ++res;
            if (target <= 0) return res;
        }
        return res;
    }
};

File: Problems/1338_Reduce_Array_Size_to_The_Half.py
from typing import List
from collections import Counter, defaultdict

class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        c = Counter(arr)
        target = (len(arr) + 1) // 2
        res = 0
        for _, cnt in c.most_common():
            target -= cnt
            res += 1
            if target <= 0:
                return res
        return res

class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        c = Counter(arr)
        target = (len(arr) + 1) // 2
        res = 0
        for _, cnt in c.most_common():
            target -= cnt
            res += 1
            if target <= 0:
                return res
        return res

class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        hm = defaultdict(int)
        target = (len(arr) + 1) // 2
        for num in arr:
            hm[num] += 1
        
        counts = list(hm.values())
        counts.sort(reverse = True)
        res = 0
        total = 0
        for cnt in counts:
            total += cnt
            res += 1
            if total >= target:
                return res
        return res